apps/src/code-studio/pd/workshop_enrollment/enroll_form.tsx

```diff
@@ -352,102 +362,129 @@ export default function EnrollForm(props: EnrollFormProps) {
         }
       );
       setIsSubmitting(false);
-      if (response.ok) {
-        const result = await response.json();
-        if (result.workshop_enrollment_status === 'error') {
-          setSubmissionErrorMessage(result.error_message || 'unknown error');
-        }
-        props.onSubmissionComplete(result);
+      const result = await response.json();
+      if (
+        result.workshop_enrollment_status === SUBMISSION_STATUSES.UNKNOWN_ERROR
+      ) {
+        setSubmissionErrorMessage(
+          result.error_message || 'Unknown error occurred'
+        );
       }
+      props.onSubmissionComplete(result);
     } catch (error) {
-      setSubmissionErrorMessage('unknown error');
+      setIsSubmitting(false);
+      setSubmissionErrorMessage('Unknown error occurred');
     }
   };
 
-  const getRequiredFieldErrors = () => {
-    let errors = getErrors();
-    const missingRequiredFields = getMissingRequiredFields();
-    const schoolInfoErrors =
-      SchoolAutocompleteDropdownWithCustomFields.validate(schoolInfoState);
+  const getAllErrors = () => {
+    const errors: FormErrors = {};
+
+    if (formState.email) {
+      if (!isEmail(formState.email)) {
+        errors.email = 'Must be a valid email address';
+      }
+      if (!props.email && formState.email !== formState.confirm_email) {
+        errors.confirm_email = 'Email addresses do not match';
+      }
+    }
 
     if (
-      missingRequiredFields.length ||
-      Object.keys(errors).length ||
-      Object.keys(schoolInfoErrors).length
+      schoolInfoInvalid({
+        country: schoolInfo.country,
+        schoolName: schoolInfo.schoolName,
+        schoolZip: schoolInfo.schoolZip,
+        schoolId: schoolInfo.schoolId,
+        schoolsList: schoolInfo.schoolsList,
+      })
     ) {
-      const requiredFieldsErrors: FormErrors = {};
+      errors.school_info = 'School information is required';
+    }
+
+    const missingRequiredFields = requiredFields.filter(
+      field =>
+        !formState[field] ||
+        (Array.isArray(formState[field]) && formState[field].length === 0)
+    );
+    if (missingRequiredFields.length) {
       missingRequiredFields.forEach(f => {
-        requiredFieldsErrors[f] = '';
+        errors[f] = 'Field is required';
       });
-      errors = {...errors, ...requiredFieldsErrors, ...schoolInfoErrors};
     }
+
     return errors;
   };
 
-  const getMissingRequiredFields = () => {
-    const requiredFields: Array<keyof EnrollFormState> = [
+  const requiredFields: Array<keyof EnrollFormState> = useMemo(() => {
+    const fields: Array<keyof EnrollFormState> = [
       'first_name',
       'last_name',
       'email',
     ];
 
     if (!props.email) {
-      requiredFields.push('confirm_email');
+      fields.push('confirm_email');
     }
 
     if (props.workshop_course === CSF) {
-      requiredFields.push('role', 'grades_teaching');
+      fields.push('role', 'grades_teaching');
```

Comment:
Can you explain what this change is doing?

Reply:
I've changed a function for getting required fields to a memoized property based on derived state. then on lines 432 to 451, I'm making the "explain" textareas required if their corresponding options are checked. this was missing from the original implementation